src/MotorPlugin.cc

```diff
@@ -0,0 +1,704 @@
+    auto result = this->registry->DeclareParameter(
+        this->scopedName, std::move(value));
+  }
+
+  //! \brief Update the parameter from the registry
+  public: void Update()
+  {
+    if (this->registry == nullptr)
+    {
+      return;
+    }
+    const double change_tolerance{1.0e-8};
+
+    auto value = std::make_unique<gz::msgs::Double>();
+    auto result = this->registry->Parameter(scopedName, *value);
+    if (result.ResultType() ==
+        gz::transport::parameters::ParameterResultType::Success)
+    {
+      const double a = this->getter(*obj);
+      const double b = value->data();
+      const bool changed = !math::equal(a, b, change_tolerance);
+      if (changed)
+      {
+        this->setter(*obj, b);
+        gzdbg << "Parameter " << this->scopedName << " updated from "
+              << a << " to " << b << std::endl;
+      }
+    }
+    else
+    {
+      gzerr << "Failed to get parameter [" << this->scopedName << "] :"
+            << result << std::endl;
+    }
+  }
+
+  private: gz::transport::parameters::ParametersRegistry * registry{nullptr};
+  private: T * obj{nullptr};
+  private: std::function<double(const T&)> getter;
+  private: std::function<void(T&, double)> setter;
+  private: std::string_view prefix;
+  private: std::string_view name;
+  private: std::string scopedName;
+};
+
+//////////////////////////////////////////////////
+class MotorPlugin::Impl
+{ 
+  //
+  public: void OnPwmMsg(const google::protobuf::Message &_msg,
+                        const gz::transport::MessageInfo &_info);
+  
+                        //! Helper to initialise and declare a PID parameter
+  public: template<typename Getter, typename Setter>
+  void DeclareParameter(
+      ParameterProxy<math::PID>& param,
+      Getter&& getter,
+      Setter&& setter,
+      const std::string& prefix)
+  {
+    param.Init(registry, &posPid,
+      std::forward<Getter>(getter),
+      std::forward<Setter>(setter),
+      prefix);
+    param.Declare();
+  }
+  /// \brief World occupied by the parent model.
+  public: World world{kNullEntity};
+  
+  /// \brief Name of the world entity.
+  public: std::string worldName;
+
+  /// \brief Model entity of Motor Model.
+  public: Model parentModel{kNullEntity};
+
+  /// \brief Name of the model entity.
+  public: std::string parentModelName;
+
+  /// \brief Array of controllers
+  public: std::vector<Control> controls;
+
+  /// \brief Array of pwm command topics
+  public: std::vector<std::string> topics;
+  
+  /// \brief keep track of controller update sim-time.
+  public: std::chrono::steady_clock::duration lastMotorModelUpdateTime{0};
+
+  ///
+  public: std::string joint_name_;
+  
+  ///
+  public: gz::sim::Entity joint_entity_ = gz::sim::kNullEntity;
+
+  /// \brief Array of subscribers for PWM topics
+  public: std::vector<gz::transport::Node::Subscriber> subscribers;
+  
+  /// \brief Store PWM values indexed by channel
+  public: std::map<std::string, std::string> pwmValues;
```

Comment:
Should not convert the pwm value from a double to a string. So if using a map should be

```c++
public: std::map<std::string, double> pwmValues;
```

However I'm not convinced we need a map either. A better way to handle multiple messages in a single plugin would be to use the `Actuator` message used in the `gz::sim::system::JointController` and `gz::sim::system::JointPositionController` plugins.